ffc/backends/ufc/dofmap.py

```diff
@@ -122,6 +122,118 @@ def tabulate_facet_dofs(L, ir):
         return L.NoOp()
 
 
+def tabulate_dof_permutations(L, ir):
+    edge_perms, facet_perms, cell = ir["dof_permutations"]
+    tdim = cell.topological_dimension()
+
+    perm = L.Symbol("perm")
+    ndofs = L.Symbol("ndofs")  # FIXME: get ndofs from somewhere
+    i = L.Symbol("i")
+    code = [L.ForRange(i, 0, ndofs, body=[L.Assign(perm[i], i)])]
+
+    if tdim == 1 or (len(edge_perms) == 0 and len(facet_perms) == 0):
+        code += [L.Return()]
+        return L.StatementList(code)
+
+    global_indices = L.Symbol("global_indices")
+
+    # Copied from ufc_geometry.h
+    edge_vertices = {'triangle': ((1, 2), (0, 2), (0, 1)),
+                     'tetrahedron': ((2, 3), (1, 3), (1, 2), (0, 3), (0, 2), (0, 1)),
+                     'quadrilateral': ((0, 1), (2, 3), (0, 2), (1, 3)),
+                     'hexahedron': ((0, 1), (2, 3), (4, 5), (6, 7), (0, 2), (1, 3),
+                                    (4, 6), (5, 7), (0, 4), (1, 5), (2, 6), (3, 7))}
+    facet_edge_vertices = {'tetrahedron': (((2, 3), (1, 3), (1, 2)),
+                                           ((2, 3), (0, 3), (0, 2)),
+                                           ((1, 3), (0, 3), (0, 1)),
+                                           ((1, 2), (0, 2), (0, 1))),
+                           'hexahedron': (((0, 1), (2, 3), (0, 2), (1, 3)),
+                                          ((4, 5), (6, 7), (4, 6), (5, 7)),
+                                          ((0, 1), (4, 5), (0, 4), (1, 5)),
+                                          ((2, 3), (6, 7), (2, 6), (3, 7)),
+                                          ((0, 2), (4, 6), (0, 4), (2, 6)),
+                                          ((1, 3), (5, 7), (1, 5), (3, 7)))}
```

Comment:
I suggest to query these table from FIAT. Here is described how to do it: https://bitbucket.org/fenics-project/ffc/issues/169/replace-tables-from-ufc_geometryh-by

Reply:
Thanks - I've now added that